Python/sysmodule.c

```diff
@@ -111,6 +112,266 @@ PySys_SetObject(const char *name, PyObject *v)
+    _Py_AuditHookEntry *e = _PyRuntime.audit_hook_head;
+    PyThreadState *ts = _PyThreadState_GET();
+    PyInterpreterState *is = ts ? ts->interp : NULL;
+    if (!e && (!is || !is->audit_hooks) && !dtrace) {
+        return 0;
+    }
+
+    PyObject *exc_type, *exc_value, *exc_tb;
+    if (ts) {
+        PyErr_Fetch(&exc_type, &exc_value, &exc_tb);
+    }
+
+    /* Initialize event args now */
+    if (argFormat && argFormat[0]) {
+        va_list args;
+        va_start(args, argFormat);
+        eventArgs = Py_VaBuildValue(argFormat, args);
+        if (eventArgs && !PyTuple_Check(eventArgs)) {
+            eventArgs = PyTuple_Pack(1, eventArgs);
+        }
+    } else {
+        eventArgs = PyTuple_New(0);
+    }
+    if (!eventArgs) {
+        goto exit;
+    }
+
+    /* Call global hooks */
+    for (; e; e = e->next) {
+        if (e->hookCFunction(event, eventArgs, e->userData) < 0) {
+            goto exit;
+        }
+    }
+
+    /* Dtrace USDT point */
+    if (dtrace) {
+        PyDTrace_AUDIT(event, (void *)eventArgs);
+    }
+
+    /* Call interpreter hooks */
+    if (is && is->audit_hooks) {
+        eventName = PyUnicode_FromString(event);
+        if (!eventName) {
+            goto exit;
+        }
+
+        hooks = PyObject_GetIter(is->audit_hooks);
+        if (!hooks) {
+            goto exit;
+        }
+
+        while ((hook = PyIter_Next(hooks)) != NULL) {
+            PyObject *ores = PyObject_CallFunctionObjArgs(
+                hook, eventName, eventArgs, NULL);
+            if (!ores) {
+                goto exit;
+            }
+            Py_DECREF(ores);
+            Py_CLEAR(hook);
+        }
+        if (PyErr_Occurred()) {
+            goto exit;
+        }
+    }
+
+    res = 0;
+
+exit:
+    Py_XDECREF(hook);
+    Py_XDECREF(hooks);
+    Py_XDECREF(eventName);
+    Py_XDECREF(eventArgs);
+
+    if (ts) {
+        if (!res) {
+            PyErr_Restore(exc_type, exc_value, exc_tb);
+        } else {
+            Py_XDECREF(exc_type);
+            Py_XDECREF(exc_value);
+            Py_XDECREF(exc_tb);
+        }
+    }
+
+    return res;
+}
+
+void _PySys_ClearAuditHooks(void) {
+    /* Must be finalizing to clear hooks */
+    _PyRuntimeState *runtime = &_PyRuntime;
+    PyThreadState *ts = _PyRuntimeState_GetThreadState(runtime);
+    assert(!ts || _Py_CURRENTLY_FINALIZING(runtime, ts));
+    if (!ts || !_Py_CURRENTLY_FINALIZING(runtime, ts))
+        return;
+
+    if (Py_VerboseFlag) {
+        PySys_WriteStderr("# clear sys.audit hooks\n");
+    }
+
+    /* Hooks can abort later hooks for this event, but cannot
+       abort the clear operation itself. */
+    PySys_Audit("sys._clearaudithooks", NULL);
+    PyErr_Clear();
+
+    _Py_AuditHookEntry *e = _PyRuntime.audit_hook_head, *n;
+    _PyRuntime.audit_hook_head = NULL;
+    while (e) {
+        n = e->next;
+        free(e);
+        e = n;
+    }
+}
+
+int
+PySys_AddAuditHook(void *hook, void *userData)
```

Comment:
This appears to be entirely distinct from `sys.addaudithook()`.  Why the distinction?

The split seems to be on "global" vs. interpreter.  Which events can happen outside an interpreter?  I'd expect them to be few, mostly relating to initialization.  For everything else, why not support adding hooks to all interpreters?  Then the audit hook machinery could be driven from the interpreters, with the pre-init events handled with one-off code.

Reply:
First there are some events that can occur outside of an interpreter, e.g. initialization or creating of a new sub-interpreter. There are also a bunch of events that will occur before a "user-space" application gets a chance to register a hook. You have to load some modules and extension classes before you get a chance to do something.

For serious business, you have to implement and register a global hook before the interpreter is initialized. The ``sys.addaudithook()`` API is of limited use for security transparency.

Reply:
It's separate because a C hook (`void*`) behaves differently from a Python hook (`PyObject*`). We have to do both.

As it stands, we have prior art here. Setting a trace func or a profile func only applies to the current thread, and you have to manually add them to any new threads. For pure-Python hooks, doing it per-interpreter is a similar constraint, and similarly useful. It needs to be documented better here (I don't think it needs to be more clearly documented in the PEP - that's not where documentation lives for CPython).